File: Easy_Leetcode_Problems/Easy_Leetcode_Problems/Relative_ranks_506.py
# Logic:
#
# 1. Sort the scores in descending order.
#
# 2. Store each score and its rank in a dictionary.
#
# 3. Assign:
#    Rank 1 -> Gold Medal
#    Rank 2 -> Silver Medal
#    Rank 3 -> Bronze Medal
#    Remaining -> Rank number.
#
# 4. Traverse the original array.
#
# 5. Replace each score with its rank.
#
# 6. Return the result.

# Algorithm:
#
# 1. Copy and sort the score array in descending order.
#
# 2. Create a dictionary.
#
# 3. Assign ranks to every score.
#
# 4. Create the answer array.
#
# 5. Traverse the original score array.
#
# 6. Store the corresponding rank.
#
# 7. Return the answer.

from typing import List

class Solution:
    def findRelativeRanks(self, score: List[int]) -> List[str]:

        sorted_scores = sorted(score, reverse=True)

        rank = {}

        for i in range(len(sorted_scores)):

            if i == 0:
                rank[sorted_scores[i]] = "Gold Medal"

            elif i == 1:
                rank[sorted_scores[i]] = "Silver Medal"

            elif i == 2:
                rank[sorted_scores[i]] = "Bronze Medal"

            else:
                rank[sorted_scores[i]] = str(i + 1)

        result = []

        for s in score:
            result.append(rank[s])

        return result
    
# Interview Explanation:
#
# 1. I sorted the scores in descending order.
#
# 2. Then I assigned medals to the top three athletes.
#
# 3. For the remaining athletes,
#    I assigned their rank numbers.
#
# 4. I stored the score-to-rank mapping in a dictionary.
#
# 5. Finally, I traversed the original array
#    and replaced each score with its rank.
#
# Time Complexity: O(n log n)
# Space Complexity: O(n)

File: Easy_Leetcode_Problems/Easy_Leetcode_Problems/Relative_ranks_506.java
// Logic:
//
// 1. Copy and sort the score array.
//
// 2. Traverse the sorted array from highest to lowest.
//
// 3. Assign:
//    - Gold Medal
//    - Silver Medal
//    - Bronze Medal
//    - Rank number
//
// 4. Store score → rank in a HashMap.
//
// 5. Traverse the original array.
//
// 6. Return the rank of each score.

// Algorithm:
//
// 1. Clone the score array.
//
// 2. Sort the cloned array.
//
// 3. Create a HashMap.
//
// 4. Traverse from the largest score.
//
// 5. Assign medals or rank numbers.
//
// 6. Create the answer array.
//
// 7. Traverse the original array.
//
// 8. Fill the answer using the HashMap.
//
// 9. Return the answer.



import java.util.*;

class Solution {
    public String[] findRelativeRanks(int[] score) {

        int[] sorted = score.clone();
        Arrays.sort(sorted);

        HashMap<Integer, String> map = new HashMap<>();

        int n = sorted.length;

        for (int i = n - 1; i >= 0; i--) {

            int rank = n - i;

            if (rank == 1) {
                map.put(sorted[i], "Gold Medal");
            } 
            else if (rank == 2) {
                map.put(sorted[i], "Silver Medal");
            } 
            else if (rank == 3) {
                map.put(sorted[i], "Bronze Medal");
            } 
            else {
                map.put(sorted[i], String.valueOf(rank));
            }
        }

        String[] answer = new String[n];

        for (int i = 0; i < n; i++) {
            answer[i] = map.get(score[i]);
        }

        return answer;
    }
}

// Interview Explanation:
//
// 1. I copied and sorted the score array.
//
// 2. Then I traversed it from the highest score
//    to the lowest score.
//
// 3. I assigned Gold, Silver and Bronze medals
//    to the top three athletes.
//
// 4. The remaining athletes received
//    their rank number.
//
// 5. I stored the score-to-rank mapping
//    in a HashMap.
//
// 6. Finally, I traversed the original array
//    and returned the ranks in the original order.
//
// Time Complexity: O(n log n)
// Space Complexity: O(n)